File: apps/web/lib/notification-view.ts
import { copy } from '@ayman/contracts/copy';
import { formatCopy } from '@ayman/contracts/format';
import type { StudentNotification } from '@ayman/contracts/notifications';
import { ASSISTANT_OPEN_PARAM } from './assistant-mount';
import { MY_BOOK_ORDERS_HREF } from './book-order-view';
import { reviewHref } from './quiz-links';

/**
 * Turning a notification row into a sentence and a destination.
 *
 * Pure functions in their own module so both the panel and the full page
 * render the same words and link to the same place — and so the mapping is
 * testable without a render. The API deliberately sends no prose (see
 * `packages/contracts/src/notifications.ts`); this is where it becomes prose.
 */

export interface NotificationView {
  title: string;
  /** A short qualifier under the title. `null` when the title says it all. */
  detail: string | null;
  /**
   * The line under the title naming WHAT this is about.
   *
   * It used to be read straight off the row as `entry.lessonTitle` by both
   * renderers, which quietly assumed every notification is about a lesson.
   * `conversation_reply` is not, and that assumption became a type error the
   * moment it landed — in two components rather than here, where the mapping
   * from row to prose already lives.
   */
  subtitle: string;
  href: string;
}

const c = copy.notifications;

export function describeNotification(entry: StudentNotification): NotificationView {
  switch (entry.kind) {
    case 'quiz_graded':
      return {
        title: formatCopy(c.quizGraded, { score: entry.scorePercent }),
        // `passed` is nullable on the wire; a missing verdict renders no
        // qualifier rather than guessing one from the score, because the pass
        // mark is per-quiz and this row does not carry it.
        detail:
          entry.passed === null ? null : entry.passed ? c.quizGradedPassed : c.quizGradedFailed,
        subtitle: entry.lessonTitle,
        href: reviewHref(entry.lessonId, entry.attemptId),
      };

    case 'extra_attempt_granted':
      return {
        title: c.extraAttempt,
        detail: null,
        subtitle: entry.lessonTitle,
        // The quiz's intro page, not a new attempt: starting a graded exam is
        // never something a link does. That page owns the button.
        href: `/quizzes/${entry.lessonId}`,
      };

    case 'conversation_reply':
      return {
        title: c.conversationReply,
        detail: null,
        subtitle: copy.assistant.title,
        /*
         * The thread lives in the widget, not on a page of its own — so this
         * links to the dashboard carrying the flag that opens it there. A
         * `/conversations/:id` route would be a second place to read the same
         * messages, and the student would have two inboxes.
         */
        href: `/dashboard?${ASSISTANT_OPEN_PARAM}=1`,
      };

    case 'instructor_message':
      return {
        // The lead-in is chosen by WHY he wrote, so a student can tell at a
        // glance whether this is their result or a reminder about the group.
        // An unknown kind — a row written by a newer deployment mid-release —
        // falls back to the generic line rather than being dropped: a message
        // from the instructor is the last thing this feed should swallow.
        title: INSTRUCTOR_LEAD_INS[entry.outreachKind] ?? c.instructorMessage,
        detail: null,
        subtitle: copy.assistant.thread.title,
        // Same destination as a reply, and for the same reason: the thread
        // lives in the widget, and a `/conversations/:id` route would give the
        // student two inboxes showing one conversation.
        href: `/dashboard?${ASSISTANT_OPEN_PARAM}=1`,
      };

    case 'payment_approved':
      return {
        title: formatCopy(c.paymentApproved, { course: entry.courseTitle }),
        detail: null,
        subtitle: entry.courseTitle,
        href: `/courses/${entry.courseSlug}`,
      };

    case 'payment_rejected':
      return {
        title: formatCopy(c.paymentRejected, { course: entry.courseTitle }),
        // The admin's own explanation, shown as the qualifier under the
        // title — same idea as `quizGradedPassed`/`quizGradedFailed`, but the
        // text is not from a fixed vocabulary this time.
        detail: entry.reason,
        subtitle: entry.courseTitle,
        href: `/courses/${entry.courseSlug}`,
      };

    case 'subscription_expiring_soon':
      return {
        title: formatCopy(c.subscriptionExpiringSoon, { course: entry.courseTitle }),
        // The exact date, formatted the same way the dashboard card's own
        // absolute-date branch does — the title already said "soon"; this is
        // where "soon" becomes a real number.
        detail: formatNotificationTime(entry.validUntil),
        subtitle: entry.courseTitle,
        // Straight to the course — same destination `payment_approved` uses,
        // and for the same reason: renewing IS opening the subscribe panel
        // from inside the course page.
        href: `/courses/${entry.courseSlug}`,
      };

    case 'subscription_cancelled':
      return {
        title: formatCopy(c.subscriptionCancelled, { course: entry.courseTitle }),
        // The admin's own explanation — same treatment as `payment_rejected`'s
        // `reason` above, and for the same underlying fact: this is free
        // text an admin chose to write, an admin chose to show, not a fixed
        // vocabulary this feed picks from.
        detail: entry.reason,
        subtitle: entry.courseTitle,
        href: `/courses/${entry.courseSlug}`,
      };

    /*
      The two ADMIN kinds, rendered by the same function as every student one.

      They land in the same feed and the same bell on purpose — an instructor
      who has to check a second place for «فيه حاجة مستنياني» checks neither.
      The hrefs point at the QUEUE rather than at the individual row: the
      decision is made in a list, beside the others waiting, and deep-linking
      to one submission hides the fact that four more arrived with it.
    */
    case 'payment_submitted':
      return {
        title: formatCopy(c.paymentSubmitted, { name: entry.studentName }),
        detail: null,
        subtitle: entry.courseTitle,
        href: '/admin/payments',
      };

    case 'book_order_placed':
      return {
        title: formatCopy(c.bookOrderPlaced, { name: entry.studentName }),
        detail: null,
        // A book order is not attached to a course — the shop sells from its
        // own catalogue — so the only thing left to name is the queue itself.
        // From `copy.notifications`, NOT the admin table: this module is
        // imported by the student's bell, and reaching into `copy/admin` here
        // would pull the whole admin copy set onto every signed-in page.
        subtitle: c.bookOrderQueue,
        href: '/admin/books',
      };

    /*
      The three STUDENT book-order kinds — `book_order_placed` above is the
      admin's alert about the same object, seen from the other side.

      All three land on `/store/orders` rather than on the shop, and that is the
      whole point of them: a student who is told «كتابك خرج ليك» and lands on a
      page selling books has been answered with an advertisement. `{book}` is
      resolved at read time off the order's first line, so a title renamed after
      shipping reads as its current name — and an order whose lines were all
      removed simply has an empty slot rather than a row that fails to parse.
    */
    case 'book_order_shipped':
      return {
        title: formatCopy(c.bookOrderShipped, { book: entry.bookTitle }),
        detail: null,
        // `copy.notifications`, NOT `copy.books.mine.title` — the two say the
        // same word today and this module is imported by the student's bell, so
        // the subtitle is kept in the same table as every other row's.
        subtitle: c.bookOrderMineQueue,
        href: MY_BOOK_ORDERS_HREF,
      };

    case 'book_order_delivered':
      return {
        title: formatCopy(c.bookOrderDelivered, { book: entry.bookTitle }),
        detail: null,
        subtitle: c.bookOrderMineQueue,
        href: MY_BOOK_ORDERS_HREF,
      };

    case 'book_order_rejected':
      return {
        title: formatCopy(c.bookOrderRejected, { book: entry.bookTitle }),
        // The admin's own words, verbatim, in the same slot `payment_rejected`
        // puts its `reason` — and for the same reason: a reason paraphrased by
        // the platform is a reason the student argues with instead of acting
        // on. The card on `/store/orders` prints it a second time under «السبب:»,
        // which is deliberate: the feed is where it is seen, that page is where
        // it stays.
        detail: entry.reason,
        subtitle: c.bookOrderMineQueue,
        href: MY_BOOK_ORDERS_HREF,
      };

    // A third ADMIN kind, same discipline as the two above.
    case 'assistant_question_received':
      return {
        title: formatCopy(c.assistantQuestionReceived, { name: entry.studentName }),
        // The snapshot of what was asked, not a fixed qualifier — same slot
        // `payment_rejected`'s `reason` occupies for the same reason: this is
        // free text, not a fixed vocabulary this feed picks from.
        detail: entry.preview || null,
        subtitle: c.assistantQuestionQueue,
        href: `/admin/inbox/${entry.conversationId}`,
      };
  }
}

/**
 * Keyed by `OUTREACH_KINDS`, but typed as an open record on purpose.
 *
 * The wire field is a bare `string` (see `notifications.ts`) because the
 * payload is jsonb and a row can outlive the build that wrote it. Typing this
 * as `Record<OutreachKind, string>` would force a cast at the lookup and buy
 * nothing — the fallback above is the real handling.
 */
const INSTRUCTOR_LEAD_INS: Record<string, string> = {
  quiz_result: c.instructorMessageQuizResult,
  quiz_nudge: c.instructorMessageQuizNudge,
  lesson_praise: c.instructorMessageLessonPraise,
  whatsapp_invite: c.instructorMessageWhatsappInvite,
};

/**
 * An ABSOLUTE timestamp, not "من ٣ ساعات".
 *
 * Relative time was the first attempt here and it was wrong twice over. It
 * needs the current clock, and reading the clock during render is impure —
 * the React Compiler rejects it outright, and rightly: a Server Component
 * rendering "من دقيقتين" and the client hydrating a moment later disagree,
 * and React discards the tree on that mismatch. Threading a `now` prop down
 * from the server works around the compiler but not the underlying problem,
 * that the string goes stale the second it is painted.
 *
 * An absolute date is also what every other timestamp in the product shows —
 * `activity-feed.tsx`, `devices-list.tsx` — with the same formatter and the
 * same Western digits (`-u-nu-latn`). One convention, and nothing to keep in
 * sync with a ticking clock.
 */
const dateFormatter = new Intl.DateTimeFormat('ar-EG-u-nu-latn', {
  dateStyle: 'medium',
  timeStyle: 'short',
});

export function formatNotificationTime(iso: string): string {
  return dateFormatter.format(new Date(iso));
}

File: apps/mobile/lib/features/notifications/domain/entities/notification_view.dart
import 'package:easy_localization/easy_localization.dart';
import 'package:flutter/material.dart';

import '../../../../core/localization/copy_keys.dart';
import '../../../../core/router/routes.dart';
import 'notification_entry.dart';

/// A notification, rendered.
@immutable
class NotificationView {
  const NotificationView({
    required this.title,
    required this.icon,
    this.detail,
    this.subtitle,
    this.route,
  });

  final String title;
  final IconData icon;

  /// A second line — a verdict, a grade, a delivery estimate.
  final String? detail;

  /// What the notification is ABOUT: a lesson title, a course name.
  final String? subtitle;

  /// Where tapping goes. Null means the row is not tappable, which is a real
  /// state — an admin-only kind reaching a student's feed has nowhere useful
  /// to send them.
  final String? route;
}

/// Turns a row into words and a destination.
///
/// ## Why the client composes the sentence at all
///
/// The API never sends prose: a row carries a `kind` plus ids and numbers, and
/// there is no `message` column on purpose. Two consequences the client must
/// honour:
///
///  1. **Titles are resolved at READ time.** A lesson renamed after the
///     notification was written reads with its NEW name. So a rendered string
///     must never be cached — re-describe from the entry every build.
///  2. **An unknown kind returns null** and the row is dropped, rather than
///     throwing. The API is on a rolling release and will one day send a kind
///     this build has not heard of.
abstract final class NotificationDescriber {
  /// Null for a kind this build does not know. Callers filter those out.
  static NotificationView? describe(NotificationEntry entry) {
    return switch (entry.kind) {
      'quiz_graded' => NotificationView(
        title: tr(
          CopyKeys.notificationsQuizGraded,
          namedArgs: {'score': '${entry.intOrNull('scorePercent') ?? 0}'},
        ),
        // `passed` is nullable in the column and should not occur. Render NO
        // verdict rather than guessing — telling a student they failed because
        // a boolean was null is the worst possible way to be wrong.
        detail: switch (entry.boolOrNull('passed')) {
          true => tr(CopyKeys.notificationsQuizGradedPassed),
          false => tr(CopyKeys.notificationsQuizGradedFailed),
          null => null,
        },
        subtitle: entry.str('lessonTitle'),
        icon: Icons.fact_check_outlined,
        route: AppRoutes.attemptReviewOf(
          entry.str('lessonId'),
          entry.str('attemptId'),
        ),
      ),

      'extra_attempt_granted' => NotificationView(
        title: tr(CopyKeys.notificationsExtraAttempt),
        subtitle: entry.str('lessonTitle'),
        icon: Icons.verified_outlined,
        // The quiz INTRO, never a fresh attempt: starting a graded exam is
        // never something a link does.
        route: AppRoutes.quizOf(entry.str('lessonId')),
      ),

      'conversation_reply' => NotificationView(
        title: tr(CopyKeys.notificationsConversationReply),
        subtitle: tr(CopyKeys.assistantThreadTitle),
        icon: Icons.forum_outlined,
        route: AppRoutes.chat,
      ),

      'instructor_message' => NotificationView(
        // Four different sentences behind one kind, chosen by `outreachKind`.
        // «شاف نتيجتك» and «فاكرك بالكويز» say different things and a single
        // generic title would waste the one line a notification gets.
        title: switch (entry.str('outreachKind')) {
          'quiz_result' => tr(CopyKeys.notificationsInstructorMessageQuizResult),
          'quiz_nudge' => tr(CopyKeys.notificationsInstructorMessageQuizNudge),
          'lesson_praise' => tr(CopyKeys.notificationsInstructorMessageLessonPraise),
          'whatsapp_invite' =>
            tr(CopyKeys.notificationsInstructorMessageWhatsappInvite),
          _ => tr(CopyKeys.notificationsInstructorMessage),
        },
        subtitle: tr(CopyKeys.assistantThreadTitle),
        icon: Icons.mark_email_unread_outlined,
        route: AppRoutes.chat,
      ),

      'payment_approved' => NotificationView(
        title: tr(
          CopyKeys.notificationsPaymentApproved,
          namedArgs: {'course': entry.str('courseTitle')},
        ),
        icon: Icons.check_circle_outline_rounded,
        route: AppRoutes.courseDetailOf(entry.str('courseSlug')),
      ),

      'payment_rejected' => NotificationView(
        title: tr(
          CopyKeys.notificationsPaymentRejected,
          namedArgs: {'course': entry.str('courseTitle')},
        ),
        icon: Icons.error_outline_rounded,
        route: AppRoutes.courseDetailOf(entry.str('courseSlug')),
      ),

      'subscription_expiring_soon' => NotificationView(
        title: tr(
          CopyKeys.notificationsSubscriptionExpiringSoon,
          namedArgs: {'course': entry.str('courseTitle')},
        ),
        icon: Icons.schedule_rounded,
        route: AppRoutes.courseDetailOf(entry.str('courseSlug')),
      ),

      'subscription_cancelled' => NotificationView(
        title: tr(
          CopyKeys.notificationsSubscriptionCancelled,
          namedArgs: {'course': entry.str('courseTitle')},
        ),
        icon: Icons.cancel_outlined,
        route: AppRoutes.courseDetailOf(entry.str('courseSlug')),
      ),

      'book_order_shipped' => NotificationView(
        title: tr(
          CopyKeys.notificationsBookOrderShipped,
          namedArgs: {'book': entry.str('bookTitle')},
        ),
        detail: tr(
          CopyKeys.notificationsBookOrderShippedDetail,
          namedArgs: {'days': '${entry.intOrNull('deliveryDays') ?? 0}'},
        ),
        subtitle: tr(CopyKeys.notificationsBookOrderMineQueue),
        icon: Icons.local_shipping_outlined,
        route: AppRoutes.storeOrders,
      ),

      'book_order_delivered' => NotificationView(
        title: tr(
          CopyKeys.notificationsBookOrderDelivered,
          namedArgs: {'book': entry.str('bookTitle')},
        ),
        subtitle: tr(CopyKeys.notificationsBookOrderMineQueue),
        icon: Icons.inventory_2_outlined,
        route: AppRoutes.storeOrders,
      ),

      'book_order_rejected' => NotificationView(
        title: tr(
          CopyKeys.notificationsBookOrderRejected,
          namedArgs: {'book': entry.str('bookTitle')},
        ),
        subtitle: tr(CopyKeys.notificationsBookOrderMineQueue),
        icon: Icons.remove_shopping_cart_outlined,
        route: AppRoutes.storeOrders,
      ),

      'course_completed' => NotificationView(
        title: tr(
          CopyKeys.notificationsCourseCompleted,
          namedArgs: {'course': entry.str('courseTitle')},
        ),
        detail: tr(CopyKeys.notificationsCourseCompletedDetail),
        icon: Icons.emoji_events_outlined,
        route: AppRoutes.courseDetailOf(entry.str('courseSlug')),
      ),

      'homework_reviewed' => _homeworkReviewed(entry),

      // ── admin kinds ────────────────────────────────────────────────────
      //
      // Rendered rather than dropped: an admin using the app is a real case
      // and these are exactly the rows they need. A student never receives
      // one — the server addresses them to `conversation:read` holders.
      'payment_submitted' => NotificationView(
        title: tr(
          CopyKeys.notificationsPaymentSubmitted,
          namedArgs: {'name': entry.str('studentName')},
        ),
        icon: Icons.payments_outlined,
        route: AppRoutes.adminPayments,
      ),

      'book_order_placed' => NotificationView(
        title: tr(
          CopyKeys.notificationsBookOrderPlaced,
          namedArgs: {'name': entry.str('studentName')},
        ),
        subtitle: tr(CopyKeys.notificationsBookOrderQueue),
        icon: Icons.shopping_bag_outlined,
        route: AppRoutes.adminBookOrders,
      ),

      'assistant_question_received' => NotificationView(
        title: tr(
          CopyKeys.notificationsAssistantQuestionReceived,
          namedArgs: {'name': entry.str('visitorName')},
        ),
        subtitle: tr(CopyKeys.notificationsAssistantQuestionQueue),
        icon: Icons.question_answer_outlined,
        route: AppRoutes.adminInbox,
      ),

      'homework_submitted' => NotificationView(
        title: tr(
          CopyKeys.notificationsHomeworkSubmitted,
          namedArgs: {'name': entry.str('studentName')},
        ),
        detail: tr(CopyKeys.notificationsHomeworkSubmittedDetail),
        icon: Icons.assignment_outlined,
        route: AppRoutes.adminHomework,
      ),

      // A kind this build has not heard of. Dropped, never thrown.
      _ => null,
    };
  }

  /// Two outcomes behind one kind, and the server drops anything else.
  ///
  /// «اتقبل ✅» carries the grade when there is one; «فيه ملاحظات» never does —
  /// a mark on a rejected submission reads as a punishment rather than as
  /// feedback, and the copy note is explicit that the wording avoids an
  /// imperative because an imperative would address a boy.
  static NotificationView _homeworkReviewed(NotificationEntry entry) {
    final accepted = entry.str('homeworkStatus') == 'accepted';
    final grade = entry.intOrNull('grade');

    return NotificationView(
      title: tr(
        accepted
            ? CopyKeys.notificationsHomeworkAccepted
            : CopyKeys.notificationsHomeworkNeedsWork,
        namedArgs: {'lesson': entry.str('lessonTitle')},
      ),
      detail: accepted && grade != null
          ? tr(CopyKeys.notificationsHomeworkAcceptedGrade, namedArgs: {'grade': '$grade'})
          : tr(
              accepted
                  ? CopyKeys.notificationsHomeworkAcceptedDetail
                  : CopyKeys.notificationsHomeworkNeedsWorkDetail,
            ),
      subtitle: entry.str('lessonTitle'),
      icon: accepted ? Icons.task_alt_rounded : Icons.rate_review_outlined,
      route: AppRoutes.lessonOf(entry.str('courseSlug'), entry.str('lessonId')),
    );
  }
}
